### PR TITLE
feat: [DEL-4228]: delegate task json support

### PR DESCRIPTION
**Pull request by abhinav-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33862)

### CODE REVIEW

File: 920-delegate-service-beans/src/main/proto/io/harness/delegate/task.proto [hunk 1]
```diff
@@ -55,6 +55,8 @@ message TaskDetails {
     // Using kryo java serialization, useful for java based task where java annotation play important role of
     // determining encrypted fields and expressions.
     bytes kryo_parameters = 3;
+    // Using json serialization - the task object in java should be serializable and have all java annotations
+    string json_parameters = 8;
```
Comment (abhinav-harness) [inline 920-delegate-service-beans/src/main/proto/io/harness/delegate/task.proto:59]: change to bytes

**Review by abhinav-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33862)* status: 11 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 2047 at r3](https://reviewable.io/reviews/harness/harness-core/33862#-N3yCiQvDgEdG6N01qij:-N3yx7MT0AFGnWqXvaFb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ddaa0dab81411d2617d8e379dd244da24b2be198/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L2047)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

we don't need this change now
</blockquote></details>

Done.
___
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 182 at r3](https://reviewable.io/reviews/harness/harness-core/33862#-N3yDHel615b9B0TBSxX:-N3yx8NZ0gcnTZdsdu4o:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ddaa0dab81411d2617d8e379dd244da24b2be198/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L182)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

consider renaming it createTaskData, as we are creating taskData object from taskDetails
</blockquote></details>

Done.
___
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 194 at r3](https://reviewable.io/reviews/harness/harness-core/33862#-N3yDkG2BD159MUruk0Z:-N3ywkbA-i4RQEc6hxff:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/ddaa0dab81411d2617d8e379dd244da24b2be198/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L194)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Unable to deserialize taskDetails
</blockquote></details>

Done
___
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 196 at r3](https://reviewable.io/reviews/harness/harness-core/33862#-N3yDw301J8DlI-qHJJw:-N3yuhg1-So1S2wN13oV:b-d32w46) ([raw file](https://github.com/harness/harness-core/blob/ddaa0dab81411d2617d8e379dd244da24b2be198/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L196)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

We can convert this else to elseif and throw exception in else

else if (taskResponseType == TaskResponseType.KRYO) {
} else {
throw new InvalidRequestException("Invalid taskResponseType", e);
}
</blockquote></details>

didnt understand this 
___
*[`950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java` line 15 at r3](https://reviewable.io/reviews/harness/harness-core/33862#-N3yFIxHDkfg0ssxVesB:-N3yslgL3ifaz5pzEpBk:b75u8b3) ([raw file](https://github.com/harness/harness-core/blob/ddaa0dab81411d2617d8e379dd244da24b2be198/950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java#L15)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

where are these changes consumed ?
</blockquote></details>

This will be consumed for json serializaton after recieving string response - io.harness.grpc.DelegateServiceGrpcImpl#getTaskData

The request object is consumed response is still not consumed as of now.



_Code quote (from 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java):_
```Java
taskData.getTaskResponseType()
```

**Comment by abhinav-harness** [comment]
trigger ti1

**Review by gaurav-jain1 — DISMISSED** [review]
> Reviewed 2 of 11 files at r1, 4 of 6 files at r2, 2 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33862)* status: :shipit: complete! ✓

___
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 196 at r4](https://reviewable.io/reviews/harness/harness-core/33862#-N42tWcK87Rwcakj1tfm:-N42tWcK87Rwcakj1tfn:b-k61exn) ([raw file](https://github.com/harness/harness-core/blob/814a13130c18a020aded0405478c496cd5dec3e2/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L196)):*
> ```Java
>         throw new InvalidRequestException("Unable to serialize params", e);
>       }
>     } else if (taskDetails.getParametersCase().equals(TaskDetails.ParametersCase.KRYO_PARAMETERS)) {
> ```

we can move this check to first place, as this is the most common case, which will avoid jump statements internally
if (kryo) {
} else if (json) {
} else {
throw exception
}



_Code quote:_
```Java
if (taskDetails.getParametersCase().equals(TaskDetails.ParametersCase.KRYO_PARAMETERS))
```

**Review by gaurav-jain1 — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33862)* status: :shipit: complete! ✓

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/TaskResponseType.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright 2022 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
+ */
+
+package io.harness.delegate.beans;
+
+public enum TaskResponseType { KRYO, JSON }
```
Comment (vistaarjuneja) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/TaskResponseType.java:10]: Nit: can we rename this to `SerializationFormat` everywhere?

**Comment by vistaarjuneja** [comment]
If you're not raising another PR, it's fine - I'll change these in my PR

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/TaskResponseType.java (hunk 2 above)
Comment (abhinav-harness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/TaskResponseType.java:10]: yes we should

**Comment by jennyjames21** [comment]
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 113 at r5](https://reviewable.io/reviews/harness/harness-core/33862#-N43nCyE2C33LfGkgOPp:-N43nCyE2C33LfGkgOPq:b234ph4) ([raw file](https://github.com/harness/harness-core/blob/c69c171851b93bd149d432937ecec0e499d03ca5/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L113)):*
> ```Java
>     this.delegateTaskServiceClassic = delegateTaskServiceClassic;
>     this.objectMapper = objectMapper;
>   }
> ```

why need to pass around add here? Can we use  dependency injection here, @Inject private ObjectMapper objectMapper;



_Code quote (from 420-delegate-service/src/main/java/io/harness/service/impl/DelegateTaskServiceImpl.java):_
```Java
 @Inject private ObjectMapper objectMapper
```

**Comment by jennyjames21** [comment]
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 186 at r5](https://reviewable.io/reviews/harness/harness-core/33862#-N43nppICrzJz1uqDk-i:-N43nppICrzJz1uqDk-j:b8wggy) ([raw file](https://github.com/harness/harness-core/blob/c69c171851b93bd149d432937ecec0e499d03ca5/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L186)):*
> ```Java
>     TaskType taskType = taskDetails.getType();
>     TaskResponseType taskResponseType;
>     if (taskDetails.getParametersCase().equals(TaskDetails.ParametersCase.KRYO_PARAMETERS)) {
> ```

Can you add kryo on else part?

**Comment by jennyjames21** [comment]
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 113 at r5](https://reviewable.io/reviews/harness/harness-core/33862#-N43nCyE2C33LfGkgOPp:-N43o9U93no0ckrHqGsJ:b32wed) ([raw file](https://github.com/harness/harness-core/blob/c69c171851b93bd149d432937ecec0e499d03ca5/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L113)):*
<details><summary><i>Previously, jennyjames21 (Jenny James) wrote…</i></summary><blockquote>

why need to pass around add here? Can we use  dependency injection here, @Inject private ObjectMapper objectMapper;
</blockquote></details>

^^

**Comment by jennyjames21** [comment]
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 213 at r5](https://reviewable.io/reviews/harness/harness-core/33862#-N43oR4B5Nasf3Kh8ebX:-N43oR4B5Nasf3Kh8ebY:b38dhn9) ([raw file](https://github.com/harness/harness-core/blob/c69c171851b93bd149d432937ecec0e499d03ca5/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L213)):*
> ```Java
>         .taskResponseType(taskResponseType)
>         .build();
>   }
> ```

please add unit test for this block of code for JSON type

**Review by jennyjames21 — COMMENTED** [review]
> Reviewed 2 of 11 files at r1, 4 of 6 files at r2, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33862)* status: 1. unresolved discussions.

**Comment by abhinav-harness** [comment]
@vistaarjuneja is taking the PR along with other go lang delegate he is working on.

**Pull request closed without merging**